[Hub Generated] Add ClientEncryptionKey management APIs in 2021-10-15-preview

anujtoshniwal opened the pull request:
Add ClientEncryptionKey management API changes to be reviewed before merging in common PR

----------------------------------------------------------------------------------------------------------------

This is a PR generated at [OpenAPI Hub](https://aka.ms/openapihub). You can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/anujtoshniwal/azure-rest-api-specs/antoshni-cosmos-db-2021-10-15-preview-cekManagement...main/).


### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [x] Other, please clarify
      Adding ClientEncryptionKey resource management APIs in the new API version.
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific langauge SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [ ] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that label “WaitForARMFeedback” will be added automatically to begin ARM API Review. Failure to comply may result in delays to the manifest.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits.

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

anujtoshniwal committed:
Adds base for updating Microsoft.DocumentDB from version preview/2021-07-01-preview to version 2021-10-15-preview

anujtoshniwal committed:
Updates readme

anujtoshniwal committed:
Updates API version in new specs and examples

anujtoshniwal committed:
ClientEncryptionKey management API changes

anujtoshniwal committed:
Address semantics validation errors

anujtoshniwal committed:
Fix representation of wrappedDek

MehaKaushik commented:
\azp run

MehaKaushik commented:
/azp run

anujtoshniwal committed:
fix Swagger prettier formatting check

abhijitpai commented:
>               "$ref": "#/definitions/ClientEncryptionKeyGetResults"

nit: singular?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:2007 in 478b637. [](commit_id = 478b637b2fa1c533f6be4d5996e098e211f72bab, deletion_comment = False)

abhijitpai commented:
>     "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DocumentDB/databaseAccounts/{accountName}/sqlDatabases/{databaseName}/clientencryptionkeys": {

nit: should the casing be clientEncryptionKeys?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:1935 in 478b637. [](commit_id = 478b637b2fa1c533f6be4d5996e098e211f72bab, deletion_comment = False)

abhijitpai commented:
>         "x-ms-long-running-operation": true,

Is this really the case?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:2020 in 478b637. [](commit_id = 478b637b2fa1c533f6be4d5996e098e211f72bab, deletion_comment = False)

abhijitpai commented:
>           "description": "Value / link of the KeyEncryptionKey."

Use reference instead of value - we don't expect the KEK itself here for security reasons.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:8165 in 478b637. [](commit_id = 478b637b2fa1c533f6be4d5996e098e211f72bab, deletion_comment = False)

abhijitpai commented:
>         "description": "Create or update a ClientEncryptionKey",

Should we add that this API is meant to be invoked via tools such as the Azure Powershell / CLI (instead of directly).

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:2019 in 478b637. [](commit_id = 478b637b2fa1c533f6be4d5996e098e211f72bab, deletion_comment = False)

abhijitpai reviewed (approved):
:shipit:

anujtoshniwal committed:
minor edits to address comments"

anujtoshniwal committed:
fix LintDiff error

wmengmsft commented:
>         }

we should omit this, right? it's not really supported here.

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:6156 in 118ac82. [](commit_id = 118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0, deletion_comment = False)

wmengmsft commented:
>           "format": "byte",

just curious, what does .NET SDK generates for this property?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:8141 in 118ac82. [](commit_id = 118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0, deletion_comment = False)

wmengmsft commented on specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/examples/CosmosDBSqlClientEncryptionKeyCreateUpdate.json:
```diff
@@ -0,0 +1,47 @@
+      "properties": {
+        "resource": {
+          "encryptionAlgorithm": "AEAD_AES_256_CBC_HMAC_SHA256",
+          "wrappedDataEncryptionKey": "This is actually an array of bytes. This request/response is being presented as a string for readability in the example",
```
>This is actually an array of bytes. This request/response is being presented as a string for readability in the example [](http://example.com/codeflow?start=39&length=119)

this is going to be a base64 encoded string, right?

anujtoshniwal replied:
This will be byte[]

anujtoshniwal commented:
> > ```
> >       "format": "byte",
> > ```
> 
> just curious, what does .NET SDK generates for this property?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:8141 in [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0). [](commit_id = [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0), deletion_comment = False)

public byte[] WrappedDataEncryptionKey { get; set; }

anujtoshniwal commented:
> > ```
> >           "$ref": "#/definitions/ClientEncryptionKeyGetResults"
> > ```
> 
> nit: singular?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:2007 in [478b637](https://github.com/Azure/azure-rest-api-specs/commit/478b637b2fa1c533f6be4d5996e098e211f72bab). [](commit_id = [478b637](https://github.com/Azure/azure-rest-api-specs/commit/478b637b2fa1c533f6be4d5996e098e211f72bab), deletion_comment = False)

This is aligned with other resources. Leaving it as is to be consistent.

anujtoshniwal commented:
> > ```
> >     }
> > ```
> 
> we should omit this, right? it's not really supported here.
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:6156 in [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0). [](commit_id = [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0), deletion_comment = False)

Not sure what you are referring to. The comment is for line 6156 which is closing bracket. 
Can you please elaborate?

Mohit-Chakraborty committed:
[Search] Rename types for better SDK code (#16039)

* Rename types for better SDK code

* Rename Speller and Captions search options

* Reorder parameters to have consistency across APIs

* Remove 'SearchIndexer' prefix from projection selectors

Remove 'global' prefix from default flags on custom entities

* Change enum name AdlsGen2->AzureDataLakeStorageGen2

Change property name storageContainer->storageContainerName

* move additional .NET SDK renames to the REST spec

* Revert changes in KnowledgeStore which has GAed

* Revert changes in CustomEntityLookupSkill which has GAed

* Revert changes in SIKSBlobProjectionSelector which has GAed

* Revert changes in SearchIndexerDataSourceType which has GAed

* Rename ignoreResetRequirements->skipIndexerResetRequirementForCache

roslynlu committed:
fix validStreamingUnits (#16143)

Co-authored-by: Roslyn Lu <roslu@microsoft.com>

anujtoshniwal committed:
remove CreateUpdateOptions as they arent applicable for CEK

anujtoshniwal commented:
> > > ```
> > >     }
> > > ```
> > 
> > 
> > we should omit this, right? it's not really supported here.
> > Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/preview/2021-10-15-preview/cosmos-db.json:6156 in [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0). [](commit_id = [118ac82](https://github.com/Azure/azure-rest-api-specs/commit/118ac8242c7d6c84c029bc7ba5c61ad4c19eeea0), deletion_comment = False)
> 
> Not sure what you are referring to. The comment is for line 6156 which is closing bracket. Can you please elaborate?

Discussed offline and removed the CreateUpdateOptions as they arent applicable here.

anujtoshniwal committed:
Merge branch 'dev-cosmos-db-Microsoft.DocumentDB-2021-10-15-preview-new-base' into antoshni-cosmos-db-2021-10-15-preview-cekManagement

leniatgh committed:
Fix schemas readme files (#16207)

* Fix schemas readme files

* fix ref

anujtoshniwal committed:
fix Model validation check

niteshvijay1995 committed:
Adds base for updating Microsoft.DocumentDB from version preview/2021-07-01-preview to version 2021-10-15-preview

niteshvijay1995 committed:
Updates readme

niteshvijay1995 committed:
Updates API version in new specs and examples

anujtoshniwal committed:
Merge branch 'dev-cosmos-db-Microsoft.DocumentDB-2021-10-15-preview-new-base' of https://github.com/niteshvijay1995/azure-rest-api-specs into antoshni-cosmos-db-2021-10-15-preview-cekManagement
